### ISSUE TITLE
Subscriber memory leak - zeromq-4.1.3

### ISSUE THREAD

**Issue by tbrandon** [zeromq/libzmq#1828]
zeromq-4.1.3

last_value_cache pub/sub using the example code.  When load testing, the subscriber appears to leak memory when the message size is 50 bytes or more.

Subscriber:
```c
#include "czmq.h"

using namespace std;

int main( int argc, char *argv [] )
{
    zctx_t *context = zctx_new();
    void *subscriber = zsocket_new( context, ZMQ_SUB );
    zsocket_connect( subscriber, "tcp://10.0.1.10:21751" );

    zsocket_set_subscribe( subscriber, "" );

    while( true ) {
        char *topic = zstr_recv( subscriber );
        if( !topic )
            break;

        zframe_t *frame = zframe_recv( subscriber );

        free( topic );
        free( frame );
    }
    zctx_destroy( &context );
    return 0;
}
```

Publisher:
```python
#!/usr/bin/env python

import sys
import time
import zmq

ip = "10.0.1.10"

def main( topic ):
    ctx = zmq.Context.instance()
    publisher = ctx.socket( zmq.PUB )
    publisher.connect( "tcp://{}:21750".format( ip ) )

    # If we don't sleep and wait for the connection, the packet is lost.
    time.sleep( 1 )
    count = 0
    while 1:
        message_data = "x" * 50
        publisher.send_multipart( ( topic, message_data, ) ) 

        count += 1
        sys.stderr.write( "publish {} {} {} bytes\n".format( count, topic, len( message_data ) ) )
        #time.sleep(0.1)


if __name__ == '__main__':
    main(sys.argv[1] if len(sys.argv) > 1 else "ch.host")
```

The publisher was run with various message sizes and sleep times.  On Ubuntu 14.04 and an embedded system, the subscriber appears to leak memory when the message size is 50 bytes or more.  At 49 bytes, no growth in memory loss was detected. 

```
$ valgrind --tool=memcheck --leak-check=full ./subscribe 
==6973== Memcheck, a memory error detector
==6973== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==6973== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==6973== Command: ./subscribe
==6973== 
^C==6973== 
==6973== HEAP SUMMARY:
==6973==     in use at exit: 8,999,484 bytes in 128,486 blocks
==6973==   total heap usage: 385,525 allocs, 257,039 frees, 19,365,362 bytes allocated
==6973== 
==6973== 140 bytes in 2 blocks are possibly lost in loss record 8 of 11
==6973==    at 0x402A17C: malloc (in /usr/lib/valgrind/vgpreload_memcheck-x86-linux.so)
==6973==    by 0x42CDAA2: zmq::msg_t::init_size(unsigned int) (msg.cpp:77)
==6973==    by 0x42F9E28: zmq::v2_decoder_t::one_byte_size_ready() (v2_decoder.cpp:92)
==6973==    by 0x42FA394: zmq::decoder_base_t<zmq::v2_decoder_t>::decode(unsigned char const*, unsigned int, unsigned int&) (decoder.hpp:136)
==6973==    by 0x42EF920: zmq::stream_engine_t::in_event() (stream_engine.cpp:317)
==6973==    by 0x42C5C48: zmq::epoll_t::loop() (epoll.cpp:176)
==6973==    by 0x42F7F1B: thread_routine (thread.cpp:96)
==6973==    by 0x4324F6F: start_thread (pthread_create.c:312)
==6973==    by 0x41EEBED: clone (clone.S:129)
==6973== 
==6973== 168 (32 direct, 136 indirect) bytes in 1 blocks are definitely lost in loss record 9 of 11
==6973==    at 0x402A9AC: operator new(unsigned int, std::nothrow_t const&) (in /usr/lib/valgrind/vgpreload_memcheck-x86-linux.so)
==6973==    by 0x42F1C52: zmq::stream_engine_t::mechanism_ready() (stream_engine.cpp:831)
==6973==    by 0x42F2107: zmq::stream_engine_t::next_handshake_command(zmq::msg_t*) (stream_engine.cpp:742)
==6973==    by 0x42EDFC2: zmq::stream_engine_t::out_event() (stream_engine.cpp:361)
==6973==    by 0x42C5C2A: zmq::epoll_t::loop() (epoll.cpp:172)
==6973==    by 0x42F7F1B: thread_routine (thread.cpp:96)
==6973==    by 0x4324F6F: start_thread (pthread_create.c:312)
==6973==    by 0x41EEBED: clone (clone.S:129)
==6973== 
==6973== 5,856 bytes in 1 blocks are definitely lost in loss record 10 of 11
==6973==    at 0x402A17C: malloc (in /usr/lib/valgrind/vgpreload_memcheck-x86-linux.so)
==6973==    by 0x42CDAA2: zmq::msg_t::init_size(unsigned int) (msg.cpp:77)
==6973==    by 0x42F9CAC: zmq::v2_decoder_t::eight_byte_size_ready() (v2_decoder.cpp:130)
==6973==    by 0x42FA394: zmq::decoder_base_t<zmq::v2_decoder_t>::decode(unsigned char const*, unsigned int, unsigned int&) (decoder.hpp:136)
==6973==    by 0x42EF920: zmq::stream_engine_t::in_event() (stream_engine.cpp:317)
==6973==    by 0x42C5C48: zmq::epoll_t::loop() (epoll.cpp:176)
==6973==    by 0x42F7F1B: thread_routine (thread.cpp:96)
==6973==    by 0x4324F6F: start_thread (pthread_create.c:312)
==6973==    by 0x41EEBED: clone (clone.S:129)
==6973== 
==6973== 8,993,250 bytes in 128,475 blocks are definitely lost in loss record 11 of 11
==6973==    at 0x402A17C: malloc (in /usr/lib/valgrind/vgpreload_memcheck-x86-linux.so)
==6973==    by 0x42CDAA2: zmq::msg_t::init_size(unsigned int) (msg.cpp:77)
==6973==    by 0x42F9E28: zmq::v2_decoder_t::one_byte_size_ready() (v2_decoder.cpp:92)
==6973==    by 0x42FA394: zmq::decoder_base_t<zmq::v2_decoder_t>::decode(unsigned char const*, unsigned int, unsigned int&) (decoder.hpp:136)
==6973==    by 0x42EF920: zmq::stream_engine_t::in_event() (stream_engine.cpp:317)
==6973==    by 0x42C5C48: zmq::epoll_t::loop() (epoll.cpp:176)
==6973==    by 0x42F7F1B: thread_routine (thread.cpp:96)
==6973==    by 0x4324F6F: start_thread (pthread_create.c:312)
==6973==    by 0x41EEBED: clone (clone.S:129)
==6973== 
==6973== LEAK SUMMARY:
==6973==    definitely lost: 8,999,138 bytes in 128,477 blocks
==6973==    indirectly lost: 136 bytes in 6 blocks
==6973==      possibly lost: 140 bytes in 2 blocks
==6973==    still reachable: 70 bytes in 1 blocks
==6973==         suppressed: 0 bytes in 0 blocks
==6973== Reachable blocks (those to which a pointer was found) are not shown.
==6973== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==6973== 
==6973== For counts of detected and suppressed errors, rerun with: -v
==6973== ERROR SUMMARY: 4 errors from 4 contexts (suppressed: 0 from 0)
```

Recommendations?

Thanks.

**Comment by bluca** [comment]
Hi,

I see your test uses CZMQ. Do you have the same result using only libzmq APIs? it would be helpful to find out, to narrow the scope.

Note that I've been running both latest dev versions of czmq and libzmq unit tests through Valgrind, and this issue never come up. So either it was fixed since 4.1, or there's no test coverage for it.

**Comment by tbrandon** [comment]
I should have used:

```c
zframe_destroy( frame );
```
Instead of the call to free. 

Thank you.
